pasta/base/annotate.py

```diff
@@ -1436,3 +1435,41 @@ def _optional_token(self, token_type, token_val):
     else:
       self.tokens.next()
       return token.src + self.ws()
+
+
+def _get_indent_diff(outer, inner):
+  """Computes the whitespace added to an indented block.
+
+  Normally, the inner indent is just the outer indent + <some whitespace>,
+  and the whitespace added is returned. However, Python2 treats tabs the same
+  as 8 spaces when determining indentation level, so the inner indent may not
+  necessarily have the outer indent as a prefix.
+
+  Arguments:
+    outer: (string) Indentation of the outer block.
+    inner: (string) Indentation of the inner block.
+  Returns:
+    The string whitespace which is added to the indentation level when moving
+    from outer to inner.
+  """
+  outer_i = iter(outer)
+  inner_i = iter(inner)
+
+  try:
+    while True:
+      try:
+        outer_c = next(outer_i)
+      except StopIteration:
+        break
+      inner_c = next(inner_i)
+      if inner_c != outer_c:
+        # Python2 treats a tab as 8 spaces
+        if sys.version_info[0] < 3:
+          spaces_i = outer_i if inner_c == '\t' else inner_i
+          if not all(' ' == next(spaces_i) for _ in range(7)):
```

Comment:
this ain't it chief :(

So, python treats tab as _moving to the next tabstop_, which is at 8 space boundaries. For example, these three indents are equivalent:

space space tab
space tab
tab

In addition, this line in particular seems wrong because it's fine for them not to match up: if the tab causes the inner to _skip_ where the outer was. So e.g. if outer = four spaces, inner = one tab, the correct answer is -- well, there's no good answer (there isn't a diff between four spaces and a tab that doesn't require mixing tabs and spaces), but probably "four spaces". The code as written would return None if I understand it correctly.

Sketching it out, I think the right algorithm should be something like:

```
def _consume_indent(current_indent, c):
  """Returns the new indent after consuming c."""
  if c == ' ':
    return current_indent + 1
  elif c == '\t':
    return 8 * (current_indent // 8 + 1)  # or whatever, not sure if math is right
  else: raise ValueError(...)

def _total_indent(indents):
  indents = 0
  for c in indents:
    indents = _consume_indent(indents, c)
  return indents

def _get_indent_diff(outer, inner):
  outer_indent = _total_indent(outer)

  inner_indent = 0
  for i, c in enumerate(inner):
    if inner_indent == outer_indent: return inner[i:]
    if inner_indent > outer_indent:
      # the inner indent used a tab to skip over the outer indentation stop.
      return ' ' * (_total_indent(inner) - outer_indent)  # mixing tabs and spaces by design??? ack
    inner_indent = _consume_indent(c)
   raise ValueError("some error message about outer being more indented than inner")
```

Test cases I think are interesting:

outer = 4 spaces, inner = tab (testing that we can skip over the outer entirely)
outer = 2 spaces and then tab, inner = 16 spaces / 2 tabs / 4 spaces and a tab and a space / etc. sky's the limit (testing tabstops)
outer = 4 spaces, inner = 2 spaces (error)
outer = 4 spaces, inner = 4 spaces (code above doesn't give error, but maybe it should)

Reply:
🤮

It's worse than I thought. But my code is buggy for what I thought the behavior was, too, oh well. Back to the drawing board!